### PR TITLE
Webpack TypeScript Client, Yarn, and Lo-Dash

### PR DESCRIPTION
**Pull request by lemueldls**
### Moved client into TypeScript #27

The former client has been refactored into TypeScript. There are still some files that aren't
completely typed out and some definite cleanup that needs to be made that will come in future
commits. When the client is ready for production level, the Workbox plugin will be added to webpack
for automatic caching as mentioned in issue #29. Next should be completing the types the server,
which I will work on once the client is done. Unless in the meanwhile someone else takes it on.

### All assets are bundled with webpack #28

Every image and audio are dynamically imported via webpack. When dynamically importing new assets be
specific as you can in the directory. i.e. instead of importing from the `../../`(`CLIENT_ROOT`),
import from `../../img/sprites/` (`CLIENT_ROOT/img/sprites/FILENAME.png`). If not, webpack will
check for all file types in all directories and cause a _huge_ memory leak. An example of dynamic
importing is as follows,

```typescript
this.filepath = (await import(`../../img/sprites/${this.id}.png`)).default;
```

### Changed primary package manager to Yarn

> When coupled together, Lerna and Yarn Workspaces can ease and optimize the management of working
> with multi-package repositories.
>
> **Yarn Workspaces** manages our dependencies. Rather than having multiple `node_modules`
> directories, it intelligently optimizes the installing of dependencies together and allows for the
> cross-linking of dependencies in a monorepo. Yarn Workspaces provide tools, like Lerna, the
> low-level primitives it needs to manage multi-package repositories.

### Changed Underscore.js to Lo-Dash

> I created Lo-Dash to provide more consistent cross-environment iteration support for arrays,
> strings, objects, and `arguments` objects. It has since become a superset of Underscore, providing
> more consistent API behavior, more features (like AMD support, deep clone, and deep merge), more
> thorough documentation and unit tests (tests which run in Node, Ringo, Rhino, Narwhal, PhantomJS,
> and browsers), better overall performance and optimizations for large arrays/object iteration, and
> more flexibility with custom builds and template pre-compilation utilities.
>
> Underscore has inconsistent support for arrays, strings, objects, and `arguments` objects. In
> newer browsers, Underscore methods ignore holes in arrays, "Objects" methods iterate `arguments`
> objects, strings are treated as array-like, and methods correctly iterate functions (ignoring
> their "prototype" property) and objects (iterating shadowed properties like "toString" and
> "valueOf"), while in older browsers they will not. Also, Underscore methods like `_.clone`
> preserve holes in arrays, while others like `_.flatten` don't.

&mdash; [John-David Dalton](https://github.com/jdalton)

### CODE REVIEW

**Review by Veradictus — APPROVED** [review]
Ya man looks good 5 me

**Comment by frycz** [comment]
Plenty of good things going on here, thumbs up! I will need a bit more time to review it.

I see that one build check has not been completed yet. 

image

@lemueldls do you know what may be the reason?

**Comment by lemueldls** [comment]
@frycz I've temporarily remove Node.js `10.x` from the CI because of this
error

```console
discord.js@12.3.1: The engine "node" is incompatible with this module. Expected version ">=12.0.0". Got "10.22.0"
```

I've also removed the test, lint, and builds as,
`test` because yarn logs an error on an empty command (`"test": ""`).
`lint` because I've set it to eslint, and the server isn't factored for that yet.
`build` because the server fails with these new TypeScript rules set.

Again, this is all a temporary fix until something like adding types to the server happens soon.

**Comment by frycz** [comment]
@lemueldls I would consider removing support for node 10.x permanently as the version won't be receiving support from 2021. 12.x and higher should be enough - @Veradictus what do you think?:



If other checks cause some issues (`test`, `lint`, `build`) it is fine to turn them off for now but we need to make them work again as soon as possible. I will create an issue for discovering possible solutions.

**Comment by lemueldls** [comment]
@frycz Agreed, Node.js `12.x` is the default version and also has Long Term Support (LTS) as of now, and we should continue to follow all versions prior, active and maintained, to the CI.

**Comment by frycz** [comment]
Follow-up issue: https://github.com/Kaetram/Kaetram-Open/issues/63

File: packages/server/ts/map/regions.ts [hunk 1]
```diff
@@ -37,10 +37,10 @@ class Regions {
     }
 
     loadDoors() {
-        let doors = map.doors;
+        const doors = map.doors;
```
Comment (frycz) [inline packages/server/ts/map/regions.ts:40]: I like it. That needs be our rule in the project - if a variable is not going to change and shouldn't change then let's use `const`. We will get a ts warning when trying to mutate it

Comment (lemueldls) [reply packages/server/ts/map/regions.ts:40]: It already is added.
https://eslint.org/docs/rules/prefer-const

**Review by Veradictus — APPROVED** [review]

**Pull request closed without merging**